test/unit/StakingPool/calculatePremium.js

```diff
@@ -0,0 +1,352 @@
+    const initialCapacityUsed = BigNumber.from(0);
+
+    const { premium } = await stakingPool.calculatePremium(
+      stakedProduct,
+      period,
+      coverAmount,
+      initialCapacityUsed,
+      totalCapacity,
+      stakedProduct.targetPrice,
+      timestamp,
+    );
+    expect(premium).to.be.equal(0);
+  });
+
+  it('should overflow when calculating premium for very large capacity', async function () {
+    const { stakingPool } = this;
+    const { timestamp } = await ethers.provider.getBlock('latest');
+    const { NXM_PER_ALLOCATION_UNIT } = this.config;
+    const stakedProduct = { ...stakedProductTemplate, nextPriceUpdateTime: timestamp };
+    const period = daysToSeconds(365);
+    const coverAmountRaw = parseEther('100');
+    const coverAmount = divCeil(coverAmountRaw, NXM_PER_ALLOCATION_UNIT);
+
+    const totalCapacity = BigNumber.from(2).pow(255);
+    const initialCapacityUsed = BigNumber.from(0);
+    const basePrice = calculateBasePrice(timestamp, stakedProduct, this.config.PRICE_CHANGE_PER_DAY);
+
+    await expect(
+      stakingPool.calculatePremium(
+        stakedProduct,
+        period,
+        coverAmount,
+        initialCapacityUsed,
+        totalCapacity,
+        stakedProduct.targetPrice,
+        timestamp,
+      ),
+    ).to.be.revertedWithPanic(17); // divide-by-zero (0x11)
+    await expect(
+      stakingPool.calculatePremiumPerYear(basePrice, coverAmount, initialCapacityUsed, totalCapacity),
+    ).to.be.revertedWithPanic(17); // divide-by-zero (0x11)
+  });
+
+  it('should calculate the premium correctly when cover amount is equal to total capacity', async function () {
+    const { stakingPool } = this;
+    const { timestamp } = await ethers.provider.getBlock('latest');
+    const { INITIAL_PRICE_DENOMINATOR, NXM_PER_ALLOCATION_UNIT } = this.config;
+    // call staking pool and calculate premium
+    const stakedProduct = { ...stakedProductTemplate, nextPriceUpdateTime: timestamp };
+    const period = daysToSeconds(365);
+    const coverAmountRaw = BigNumber.from(2).pow(96);
+    const coverAmount = divCeil(coverAmountRaw, NXM_PER_ALLOCATION_UNIT);
+    const initialCapacityUsed = 0;
+    const totalCapacity = coverAmount;
+    const basePrice = calculateBasePrice(timestamp, stakedProduct, this.config.PRICE_CHANGE_PER_DAY);
+    const { premium: expectedPremium } = await stakingPool.calculatePremium(
+      stakedProduct,
+      period,
+      coverAmount,
+      initialCapacityUsed,
+      totalCapacity,
+      stakedProduct.targetPrice,
+      timestamp,
+    );
+    const premiumPerYear = await stakingPool.calculatePremiumPerYear(
+      basePrice,
+      coverAmount,
+      initialCapacityUsed,
+      totalCapacity,
+    );
+    const { surgePremium, surgePremiumSkipped } = calculateSurgePremiums(
+      coverAmountRaw,
+      initialCapacityUsed,
+      totalCapacity,
+      this.config,
+    );
+    // Note must use rounded down value to match precision loss in contracts
```

Comment:
Where exactly are we rounding down?

Reply:
The comment is confusing, so I removed it. I was referring to the fact that the `basePremium` calculation has to use `coverAmount.divCeil(NXM_PER_ALLOCATION_UNITS)` and then multiply that by `NXM_PER_ALLOCATION_UNITS` instead of using the original cover amount.